src/inmanta/server/services/compilerservice.py

```diff
@@ -248,62 +249,53 @@ async def ensure_compiler_venv(self) -> None:
-            except FileNotFoundError:
-                # Broken link or does not exist
-                await self._info(f"Creating new venv at {versioned_venv_dir_full}")
-                os.unlink(venv_dir)
-                do_create = True
-        elif not os.path.exists(venv_dir):
-            # Doesn't exist
-            do_create = True
-            await self._info(f"Creating new venv at {versioned_venv_dir_full}")
-        else:
-            # We have an older (< iso8) setup without symlink
-            virtual_env = VirtualEnv(venv_dir)
-            if virtual_env.exists():
-                try:
-                    virtual_env.can_activate()  # raises exception
-                    # version matches, move it to the correct folder
-                    os.rename(venv_dir, versioned_venv_dir_full)
-                    await self._info(f"Moving existing venv from {venv_dir} to {versioned_venv_dir_full}")
-                except VenvActivationFailedError:
-                    # Version doesn't match, move to alternative location
-                    os.rename(venv_dir, venv_dir + "_old")
-                    await self._info(f"Discarding existing venv from {venv_dir} to {venv_dir}_old, Creating new")
-                    do_create = True
-            else:
-                # otherwise broken
+        async def ensure_venv() -> None:
+            """
+            Ensures that a compatible venv exists at .venv-py<version>
+            """
+            if os.path.exists(versioned_venv_dir_full):
+                return
+            # migration from old .env
+            if os.path.exists(venv_dir) and not os.path.islink(venv_dir):
+                virtual_env = VirtualEnv(venv_dir)
+                if virtual_env.exists():
+                    with contextlib.suppress(VenvActivationFailedError):
+                        virtual_env.can_activate()  # raises exception
+                        # version matches, move it to the correct folder
+                        os.rename(venv_dir, versioned_venv_dir_full)
+                        await self._info(f"Moving existing venv from {venv_dir} to {versioned_venv_dir_full}")
+                        # All done
+                        return
+                # version doesn't match
                 os.rename(venv_dir, venv_dir + "_old")
                 await self._info(f"Discarding existing venv from {venv_dir} to {venv_dir}_old, Creating new")
-                do_create = True
 
-        if do_create:
+            # No there yet
+            await self._info(f"Creating new venv at {versioned_venv_dir_full}")
             virtual_env = VirtualEnv(versioned_venv_dir_full)
             virtual_env.init_env()
 
-        # Relative symlink is a bit exotic in python
-        try:
-            dir_fd = os.open(project_dir, os.O_RDONLY)
-            os.symlink(versioned_venv_dir, ".env", dir_fd=dir_fd)
-        finally:
-            if dir_fd:
-                os.close(dir_fd)
+        await ensure_venv()
```

Comment:
Only a style preference, but this is hidden a bit squeezed in between the two method definitions. I'd prefer the two calls at the bottom, so you have definitions of behavior first, followed by the "script".